src/Space.cpp

```diff
@@ -1002,23 +1002,40 @@ void Space::UpdateBodies()
 	m_processingFinalizationQueue = true;
 #endif
 
-	for (Body *rmb : m_removeBodies) {
-		rmb->SetFrame(FrameId::Invalid);
-		for (Body *b : m_bodies)
-			b->NotifyRemoved(rmb);
-		if (Pi::GetView()) Pi::game->GetSystemView()->BodyInaccessible(rmb);
-		m_bodies.remove(rmb);
-	}
-	m_removeBodies.clear();
-
-	for (Body *killb : m_killBodies) {
-		for (Body *b : m_bodies)
-			b->NotifyRemoved(killb);
-		if (Pi::GetView()) Pi::game->GetSystemView()->BodyInaccessible(killb);
-		m_bodies.remove(killb);
-		delete killb;
+	// removing or deleting bodies from space in one pass
+	// using integer index instead of iterator, because pop_back() invalidates the iterator to the last element
+	for (uint32_t i = 0; i < m_bodies.size();) {
```

Comment:
I'm not sure this algorithm is optimal. The previous version was O(M * N), whereas this one seems like it's O(M * N log N), degrading to O(M log M * N<sup>2</sup>). This discrepancy mostly comes from the need to iterate the main bodies list, then do a linear search (protip: `std::binary_search` is useful for logarithmic-time searches of sorted vectors) of, in the worst case, both the removed and killed bodies list, and then do another full iteration of the main bodies list.

I think a much better way to do it is to store a `uint32_t` flag in a single list of bodies to remove, iterate the removed list once to notify bodies and delete, and then clear the removed list. That should (technically) be the most efficient method short of maintaining a separate list of bodies to receive delete notifications; which would drop it to O(M * log N).

Reply:
@Web-eWorks  it seems to me that since the notification loop is performed only once for object to be deleted, (after `std::find`, not during), the complexity of this algorithm is O(N * M + M * N) -> O(N * M) ?

Reply:
You do have a point, it's closer to O(N + M * N) than it is to O(M * N log N). It's still a more expensive version of the algorithm especially at a lower number of deleted bodies due to needing an additional loop over all elements in the bodies list and then a linear search in all containers of deleted bodies, whereas the prior version at least didn't incur two `std::find`s for every live body in the space.

Don't get me wrong, I appreciate the work you've gone to in implementing this! I'd just like to be sure we're using the most efficient and effective algorithms when the PR is intended to use a more efficient and effective data container...